ENH: Add csd exercise

kaitj opened the pull request:
~~Resolves~~ Relates to #30 - adds an exercise to simulate crossing angles of 90, 60, 45, 30, and 20 degrees. I did have run into some issues simulating with `default_sphere` and ended up using `repulsion724` in the solutions. If someone can double check that, that would be great. I've also left the exercise out of the jupyter notebooks for now. Also have a figure saved showing the solutions at each crossing angle, however can't seem to get it to load in the solutions. If anyone knows how, the file is named `odf_multiple_angles.png` in the CSD figure directory.

We can add more exercises later on and also increase the difficulty, but I think this is a good place to start to see where CSD may start to fail. 

Another exercise mentioned would be to actually perform the deconvolution, but I think we need to dig a little deeper into this dataset before that. I think as we expand the episode to include more details spherical harmonics and figure out some of the quirks we are seeing with the dataset, we can add an exercise to deconvolve with different harmonic orders.

josephmje commented on _episodes/constrained_spherical_deconvolution.md:
```diff
@@ -425,5 +425,76 @@ References
+> > 
+> > plt.show()
+> > ~~~
+> > {: .language-python}s
```
```suggestion
> > {: .language-python}
```

jhlegarreta commented on _episodes/constrained_spherical_deconvolution.md:
```diff
@@ -425,5 +425,76 @@ References
+> > from dipy.viz import window, actor
+> > from dipy.sims.voxel import multi_tensor_odf
+> >
+> > # Create output directory to stoage image
```
_Create output directory to stoage image_ : _Create the output directory to store the image_

kaitj replied:
This should be fixed in dc3644f / d270e8f

jhlegarreta replied:
Looks like it has not been fixed.

kaitj replied:
Weird - will take a closer look when updating this entire episode next week.

kaitj replied:
Should actually be fixed in e1e0e2e

jhlegarreta reviewed (commented):
Have not tried c&p-ing the code into a script/notebook to try.

IMHO: the exercise statement is misleading:
> In this episode, the ODF of a single fibre population (i.e. no crossing fibres) was visualized.

I'd say that it is inaccurate; the reconstructions, if visualized with enough detail, would show crossing fibers.

If this was meant to refer to the FRF, then we run the risk of generating another confusion by offering learners an exercise that creates fODFs of crossing fibers, which some may think that could be used as a sort of advanced FRF.

So I'd restate the exercise statement.

> the file is named odf_multiple_angles.png in the CSD figure directory.

No figure has been added with the commit. Since the jupyter notebook part is still missing, it can done when adding those.

Marking the PR as a draft might be useful until it is complete.

Also, I wouldn't mark #30 as resolved; in #30 it was suggested (unless the wording used was inaccurate or confusing) to perform the actual deconvolution given the peaks for crossing fibers.

> Another exercise mentioned would be to actually perform the deconvolution, but I think we need to dig a little deeper into this dataset before that. I think as we expand the episode to include more details spherical harmonics and figure out some of the quirks we are seeing with the dataset, we can add an exercise to deconvolve with different harmonic orders.

Reading this maybe #30 was unclear: it implied using synthetic data (a voxel) (or voxels from the data containing crossings at different angles might do the job) to simulate a signal containing crossings at different angles. I'll add that to the issue.

But I agree, either solution requires more time to investigate and develop in case of the synthetic signal.

Given the time limitations, we may want to go with what this PR proposes.

kaitj commented:
> I'd say that it is inaccurate; the reconstructions, if visualized with enough detail, would show crossing fibers.
Fair point, I've restated the sentence to note in the episode we visualized the response function's ODF.

> No figure has been added with the commit. Since the jupyter notebook part is still missing, it can done when adding those.
You're right - committed, but forgot to push. Figure should be present now.

>Also, I wouldn't mark #30 as resolved
Updated the initial PR so it mentions #30 without linking to the issue.

jhlegarreta commented on _episodes/constrained_spherical_deconvolution.md:
```diff
@@ -425,5 +425,82 @@ References
+> ## Exercise 1
+>
+> In this episode, the response function was estimated from a region of
+> single coherent fibre populations and its ODF was visualized. Alternatively, 
```
`In this episode, the response function was estimated from a region of single coherent fibre populations (...)` : gives the impression that it is a specificity of the episode. Is there any other sensible way to find FRFs from real data?

`Alternatively, (...)`: adds to the confusion above IMHO.

kaitj replied:
Just removed this first part of the exercise in 9434a10 - it will need to be worked later on, but the simulation is enough until we can work out some of the issues we are seeing in the dataset.

jhlegarreta commented on _episodes/constrained_spherical_deconvolution.md:
```diff
@@ -425,5 +425,82 @@ References
+>
+> In this episode, the response function was estimated from a region of
+> single coherent fibre populations and its ODF was visualized. Alternatively, 
+> We can simulate what the ODF would look like for multiple fibre
```
We: do not capitalize

kaitj replied:
Same comment as previous.

jhlegarreta reviewed (commented):
Not adding the changes to the notebooks in this PR increases again the chances of going out-of-sync across the markdowns and notebooks.

jhlegarreta commented:
Please consider merging #123, #124 and #125 before. Thanks.

kaitj commented:
Similar to #120 - going to change this to draft now with plans to sit down one of the nights next week to just solely update this episode with figures, sync content between markdown and notebook(s), run a linter through the notebook and address the changes..

kaitj commented:
Notebooks and markdown should be synced. Added a solutions notebook for the exercise, and removed the `skip nbval` tags in the non-solutions notebook. Actions workflow has been updated to reflect this. Figures also updated after re-running through notebooks.

jhlegarreta reviewed (approved):
Thanks Jason.
